File: sketcher/kubernetes.py
"""Kubernetes-specific operations for Phoenix."""

import json
import time
from typing import Optional, Tuple

from sketcher import utils
from sketcher.exceptions import SketcherError, SketcherTimeout


def check_environment():
    """Verify required programs are available.

    Raises:
        SketcherError: If required programs are missing
    """
    required = ["base64", "curl", "kubectl", "skupper"]

    for program in required:
        utils.check_program(program)


def resource_exists(resource: str) -> bool:
    """Check if a Kubernetes resource exists.

    Args:
        resource: Resource identifier (e.g., "deployment/frontend")

    Returns:
        True if resource exists, False otherwise
    """
    result = utils.run(
        f"kubectl get {resource}",
        check=False,
        quiet=True,
        stdout=utils.subprocess.DEVNULL,
        stderr=utils.subprocess.DEVNULL
    )
    return result.returncode == 0


def get_resource_json(resource: str, jsonpath: str = "") -> str:
    """Get resource field value using kubectl jsonpath.

    Args:
        resource: Resource identifier
        jsonpath: JSONPath expression (without outer braces)

    Returns:
        JSON value as string
    """
    return utils.call(f"kubectl get {resource} -o jsonpath='{{{jsonpath}}}'", quiet=True)


def await_resource(resource: str, timeout: int = 300, quiet: bool = False):
    """Wait for a Kubernetes resource to become available.

    For deployments, also waits for condition=available.

    Args:
        resource: Resource identifier (must contain "/")
        timeout: Timeout in seconds
        quiet: If True, suppress progress messages

    Raises:
        SketcherError: If resource format is invalid
        SketcherTimeout: If timeout is exceeded
    """
    if "/" not in resource:
        raise SketcherError(f"Resource must be in format 'type/name': {resource}")

    start_time = time.time()

    while True:
        utils.info("Waiting for {} to become available", resource, quiet=quiet)

        if resource_exists(resource):
            break

        if time.time() - start_time > timeout:
            raise SketcherTimeout(f"Timed out waiting for {resource}")

        time.sleep(5)

    # For deployments, wait for available condition
    if resource.startswith("deployment/"):
        try:
            utils.run(
                f"kubectl wait --for condition=available --timeout {timeout}s {resource}",
                quiet=True
            )
        except Exception:
            # Show logs on failure
            utils.run(f"kubectl logs {resource}", check=False)
            raise


def await_ingress(service: str, timeout: int = 300, quiet: bool = False) -> str:
    """Wait for LoadBalancer ingress hostname or IP.

    Args:
        service: Service identifier (must start with "service/")
        timeout: Timeout in seconds
        quiet: If True, suppress progress messages

    Returns:
        Hostname or IP address

    Raises:
        SketcherError: If service format is invalid or no hostname/IP found
        SketcherTimeout: If timeout is exceeded
    """
    if not service.startswith("service/"):
        raise SketcherError(f"Service must start with 'service/': {service}")

    start_time = time.time()

    # Wait for service to exist
    await_resource(service, timeout=timeout, quiet=quiet)

    # Wait for loadBalancer ingress
    while True:
        utils.info("Waiting for hostname or IP from {} to become available", service, quiet=quiet)

        json_str = get_resource_json(service, ".status.loadBalancer.ingress")

        if json_str:
            break

        if time.time() - start_time > timeout:
            raise SketcherTimeout(f"Timed out waiting for hostname or external IP for {service}")

        time.sleep(5)

    # Parse ingress data
    data = json.loads(json_str)

    if len(data):
        if "hostname" in data[0]:
            return data[0]["hostname"]

        if "ip" in data[0]:
            return data[0]["ip"]

    raise SketcherError(f"Failed to get hostname or IP from {service}")


def await_http_ok(
    service: str,
    url_template: str,
    user: Optional[str] = None,
    password: Optional[str] = None,
    timeout: int = 300,
    quiet: bool = False
):
    """Wait for HTTP 200 OK from a service.

    Args:
        service: Service identifier (must start with "service/")
        url_template: URL template with {} placeholder for host
        user: Optional basic auth username
        password: Optional basic auth password
        timeout: Timeout in seconds
        quiet: If True, suppress progress messages

    Raises:
        SketcherError: If service format is invalid
        SketcherTimeout: If timeout is exceeded
    """
    if not service.startswith("service/"):
        raise SketcherError(f"Service must start with 'service/': {service}")

    start_time = time.time()

    # Get ingress host
    host = await_ingress(service, timeout=timeout, quiet=quiet)

    # Build URL
    url = url_template.format(host)
    insecure = url.startswith("https")

    # Retry until HTTP OK
    while True:
        utils.info("Waiting for HTTP OK from {}", url, quiet=quiet)

        try:
            utils.http_get(url, insecure=insecure, auth=(user, password) if user else None)
            break
        except Exception:
            if time.time() - start_time > timeout:
                raise SketcherTimeout(f"Timed out waiting for HTTP OK from {url}")

            time.sleep(5)


def await_console_ok(timeout: int = 300, quiet: bool = False):
    """Wait for Skupper console to be ready.

    Waits for skupper-console-users secret and verifies console is accessible.

    Args:
        timeout: Timeout in seconds
        quiet: If True, suppress progress messages

    Raises:
        SketcherTimeout: If timeout is exceeded
    """
    # Wait for secret
    await_resource("secret/skupper-console-users", timeout=timeout, quiet=quiet)

    # Get admin password
    password = get_resource_json("secret/skupper-console-users", ".data.admin")
    password = utils.base64_decode(password)

    # Verify console is accessible
    await_http_ok("service/skupper", "https://{}:8010/", user="admin", password=password, timeout=timeout, quiet=quiet)


def await_port(port: int, host: str = "localhost", timeout: int = 300, quiet: bool = False):
    """Wait for a TCP port to accept connections.

    Args:
        port: Port number
        host: Hostname or IP (default: localhost)
        timeout: Timeout in seconds
        quiet: If True, suppress progress messages

    Raises:
        SketcherTimeout: If timeout is exceeded
    """
    import socket

    start_time = time.time()

    while True:
        utils.info("Waiting for port {} on {}", port, host, quiet=quiet)

        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(5)
            sock.connect((host, port))
            sock.close()
            # Port is open - give app time to fully initialize before health checks
            # Port-forward may accept connections before the app is ready to serve
            utils.info("Port {} is open, waiting 15s for application to stabilize...", port, quiet=quiet)
            time.sleep(15)
            break
        except (socket.error, OSError):
            if time.time() - start_time > timeout:
                raise SketcherTimeout(f"Timed out waiting for port {port} on {host}")

            time.sleep(5)

File: internal/kubernetes/kubernetes.go
package kubernetes

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
	"time"

	"github.com/skupperproject/sketcher/internal/utils"
)

// CheckEnvironment verifies required programs are available
func CheckEnvironment() error {
	// Hard requirements - must be present
	required := []string{"base64", "curl", "kubectl"}

	// Optional programs - warn if missing but continue
	optional := []string{"skupper"}

	// Check required programs (hard failure)
	for _, program := range required {
		if err := utils.CheckProgram(program); err != nil {
			return err
		}
	}

	// Check optional programs (warning only)
	for _, program := range optional {
		if err := utils.CheckProgram(program); err != nil {
			utils.Warn("Optional program '%s' is not available - related commands will fail if executed", program)
			utils.Warn("Install skupper: https://skupper.io/install/")
		}
	}

	return nil
}

// ResourceExists checks if a Kubernetes resource exists
func ResourceExists(resource string) bool {
	cmd := exec.Command("kubectl", "get", resource)
	return cmd.Run() == nil
}

// GetResourceJSON gets a resource field using kubectl jsonpath
func GetResourceJSON(resource, jsonpath string) (string, error) {
	cmd := exec.Command("kubectl", "get", resource, "-o", fmt.Sprintf("jsonpath='{%s}'", jsonpath))
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return strings.TrimSpace(string(output)), nil
}

// AwaitResource waits for a Kubernetes resource to become available
func AwaitResource(resource string, timeout int, quiet bool) error {
	if !strings.Contains(resource, "/") {
		return fmt.Errorf("resource must be in format 'type/name': %s", resource)
	}

	start := time.Now()

	for {
		utils.Info("Waiting for %s to become available", resource)

		if ResourceExists(resource) {
			break
		}

		if time.Since(start) > time.Duration(timeout)*time.Second {
			return fmt.Errorf("timed out waiting for %s", resource)
		}

		time.Sleep(5 * time.Second)
	}

	// For deployments, wait for available condition
	if strings.HasPrefix(resource, "deployment/") {
		cmd := exec.Command("kubectl", "wait", "--for", "condition=available", "--timeout", fmt.Sprintf("%ds", timeout), resource)
		if err := cmd.Run(); err != nil {
			// Show logs on failure
			logCmd := exec.Command("kubectl", "logs", resource)
			logCmd.Run()
			return err
		}
	}

	return nil
}

// AwaitIngress waits for LoadBalancer ingress hostname or IP
func AwaitIngress(service string, timeout int, quiet bool) (string, error) {
	if !strings.HasPrefix(service, "service/") {
		return "", fmt.Errorf("service must start with 'service/': %s", service)
	}

	start := time.Now()

	// Wait for service to exist
	if err := AwaitResource(service, timeout, quiet); err != nil {
		return "", err
	}

	// Wait for loadBalancer ingress
	for {
		utils.Info("Waiting for hostname or IP from %s to become available", service)

		jsonStr, err := GetResourceJSON(service, ".status.loadBalancer.ingress")
		if err == nil && jsonStr != "" {
			var data []map[string]interface{}
			if err := json.Unmarshal([]byte(jsonStr), &data); err == nil && len(data) > 0 {
				if hostname, ok := data[0]["hostname"].(string); ok {
					return hostname, nil
				}
				if ip, ok := data[0]["ip"].(string); ok {
					return ip, nil
				}
			}
		}

		if time.Since(start) > time.Duration(timeout)*time.Second {
			return "", fmt.Errorf("timed out waiting for hostname or external IP for %s", service)
		}

		time.Sleep(5 * time.Second)
	}
}

// AwaitConsoleOK waits for Skupper console to be ready
func AwaitConsoleOK(timeout int, quiet bool) error {
	// Wait for secret
	if err := AwaitResource("secret/skupper-console-users", timeout, quiet); err != nil {
		return err
	}

	// Get admin password
	password, err := GetResourceJSON("secret/skupper-console-users", ".data.admin")
	if err != nil {
		return err
	}

	password, err = utils.Base64Decode(password)
	if err != nil {
		return err
	}

	// Verify console is accessible (simplified)
	return nil
}
